protos/tunneloffload.proto

```diff
@@ -0,0 +1,352 @@
+// Add licensing
+
+syntax = "proto3";
+
+package tunneloffload.v1;
+
+//
+// Errors for adding a session
+// If all sessions are successful inserted return _ACCEPTED
+//
+// If check of session capacity in offload device is insufficient to add all sessions 
+// do not insert any sessions and return  _REJECTED_SESSION_TABLE_FULL. It is the 
+// responsibility of the client to re-try
+//
+// If the server is unavailable for some other reason then return _REJECTED_SESSION_TABLE_UNAVAILABLE.
+// It is the  responsibility of the client to re-try
+//
+// All other errors will return _REJECTED with a buit mask of the failed sessions and it is the responsibility
+// of the client to address the issues
+//
+// ADD_SESSION_STATUS Codes Description
+//
+// _SESSION_ACCEPTED: Session is accepted by the server and the client performs normal operation
+// _SESSION_REJECTED: Session is rejected by the server as the message 
+//    is invalid, the client needs to correct the error.
+// _SESSION_TABLE_FULL: Session is rejected by the server as its session table is full, 
+//    the client needs to backoff until more space is available
+// _SESSION_TABLE_UNAVAILABLE: Session is rejected by the server due to an internal error 
+//    in the server, the client needs to back off until error is corrected.
+// _SESSION_ALREADY_EXISTS: Session is rejected by the the server as it already exists 
+//    in the server session table, the client will take corrective action to ensure state is consistent.
+//
+enum ADD_TUNNEL_STATUS {
+  _SESSION_ACCEPTED = 0;
+  _SESSION_REJECTED = 1;
+  _SESSION_TABLE_FULL = 2;
+  _SESSION_TABLE_UNAVAILABLE = 3;
+  _SESSION_ALREADY_EXISTS= 4;
+}
+
+service ipTunnelService {
+  // Get which capabilities are available while using the  
+  rpc Capabilities(CapabilityRequest) returns (CapabilityResponse);
+
+  // Creation of IP Tunnel
+  // This API shuold be generic and allow creations of many IP tunnels
+  rpc createIpTunnel(stream ipTunnelRule) returns (ipTunnelResponses) {}
+
+  // Updating the tunnel,
+  // Only the tunnelId should be constant, all other parameters can be changed in the tunnel
+  rpc updateIpTunnel(stream ipTunnelRule) returns (ipTunnelResponses) {}
```

Comment:
I t's not clear where the tunnelId is in the ipTunnelRule (renamed updateIpTunnelRequest so we can have a different structure for update is needed but if we use the same structure as createIpTunnel then we need to include the tunnelId outside of the Match criteria